### ISSUE TITLE
Stability annotation on defaulted generics fails

### ISSUE THREAD

**Issue by TimDiekmann** [rust-lang/rust#79499]
I'm adding `<A: AllocRef = Global>` to string types. Strings don't have a generic parameter, so the defaulted generic parameter is the first parameter:

```rust
pub struct String<#[unstable(feature = "allocator_api", issue = "32838")] A: AllocRef = Global> {}
```

However, the compiler complains, that this annotation is "useless":

```text
error: This stability annotation is useless
   --> library/alloc/src/string.rs:280:75
    |
280 | pub struct String<#[unstable(feature = "allocator_api", issue = "32838")] A: AllocRef = Global> {
    |                                                                           ^
```

Was this simply not added by #77118 or is this a false positive in general?

cc @varkor @Avi-D-coder

**Comment by varkor** [comment]
The error is occurring here, for reference:
https://github.com/rust-lang/rust/blob/1d2726726f8f3128e98191e4c6cb94bd76d0ddd4/compiler/rustc_passes/src/stability.rs#L156-L161
Off the top of my head, I'm not sure which part of the condition is satisfied. There's no reason I see to expect this not to work, so it could well have been a case overlooked in #77118.

@Avi-D-coder: would you be interested in taking a look to see what's going wrong? I should be able to give more immediate response than last time (it'd just take me a little while to find time to debug, myself) :)

**Comment by zachs18** [comment]
If I remove `derive(PartialOrd, Eq, Ord)` from `struct String` (i.e. manually implement them, like `PartialEq` already was), then I don't get the `This stability annotation is useless` error anymore, so I think the error might be coming from a derive macro instead of the struct declaration itself.

---

Specifically, for this code (outside of stdlib, because IDK how to `cargo expand` inside stdlib)
```rs
#![feature(allocator_api)]
#![feature(std_internals)]

use std::alloc::{Allocator, Global};

#[derive(PartialOrd, Eq, Ord)]
pub struct String<#[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    vec: Vec<u8, A>,
}
impl<A: Allocator, B: Allocator> PartialEq<String<B>> for String<A> {
    fn eq(&self, other: &String<B>) -> bool { todo!() }
}
```
`cargo +nightly expand` gives
```rs
// error "stability attributes may not be used outside of the standard library" ignored
// .. etc
impl<#[unstable(feature = "allocator_api", issue = "32838")] A: Allocator>
    ::core::marker::StructuralEq for String<A>
{
}
// .. etc (None of the other impls have the stability annotation)
```
and I think this `impl` is what's causing the error; it just looks like its the struct definition because the `Eq` derive macro copies the span from the struct definition.

---

If this is correct, maybe this issue should be renamed to something like "`Eq` derive macro copies stability annotation on generics into `impl StructuralEq`, causing compilation error"?

(Either way, if `String` (/`CString`/etc) is made allocator-aware, it would be better to manually implement `(Partial){Eq,Ord}` anyway, so as not to have an `<A: Allocator + (Partial){Eq,Ord}>` bound.)

**Comment by SaadiSave** [comment]
If work here has stalled, I'd like to make progress on this issue
